src/common/yaml.hpp

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_YAML_HPP
+#define MENDER_COMMON_YAML_HPP
+
+#include <common/config.h>
+
+#include <map>
+#include <string>
+#include <unordered_map>
+
+#include <common/common.hpp>
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/io.hpp>
+
+#ifdef MENDER_USE_YAML_CPP
+#include <yaml-cpp/yaml.h>
+#endif
+
+namespace mender {
+namespace common {
+namespace yaml {
+
+using namespace std;
+
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+namespace common = mender::common;
+
+enum YamlErrorCode {
+	NoError = 0,
+	ParseError,
+	KeyError,
+	IndexError,
+	TypeError,
+};
+
+class YamlErrorCategoryClass : public error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+extern const YamlErrorCategoryClass YamlErrorCategory;
+
+error::Error MakeError(YamlErrorCode code, const string &msg);
+
+using ExpectedString = mender::common::expected::ExpectedString;
+using ExpectedInt64 = mender::common::expected::ExpectedInt64;
+using ExpectedDouble = mender::common::expected::ExpectedDouble;
+using ExpectedBool = mender::common::expected::ExpectedBool;
+using ExpectedSize = mender::common::expected::ExpectedSize;
+
+class Yaml {
+public:
+	using ExpectedYaml = expected::expected<Yaml, error::Error>;
+	using ChildrenMap = map<string, Yaml>;
+	using ExpectedChildrenMap = expected::expected<ChildrenMap, error::Error>;
+
+	Yaml() = default;
+
+	Yaml(const Yaml &) = default;
+	Yaml &operator=(Yaml &) = default;
```

Comment:
Is this constructor necessary..? AFAICT, this constructor will never be used, because you can always convert the argument to const. OTOH, picking constructors is where C++ really is an incomprehensible mess, so maybe I'm missing something....